Support opening file paths given relative to home directory with tildes in table module

svank opened the pull request:
### Description
This PR is split from #13107 and addresses #12778 within the `table` module. The changes allow file paths to be given relative to the user's home directory as `~/dir/filename.fits` when doing file IO with `Table`.

While this PR is for the `Table` API, the actual changes do occur in `io.registry`.

For testing, I used pytest's monkeypatch to temporarily update the environment variables so that `~` resolves to the test's temp directory, only for designated tests. Then I added such a test that exercises the `write` and `read` methods with tilde-prefixed paths. In `io.misc` I also added a test that exercises reading and writing HDF5 files through `Table` with tilde-prefixed paths, to ensure we're interoperating with `h5py` correctly.

Since `Table` sometimes wraps `io.ascii`, I'll note that related changes to `io.ascii` are in #13130, though those changes and tests seem to stand independently from these changes and tests.

### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

pllim commented on astropy/table/tests/test_table.py:
```diff
@@ -3023,3 +3038,14 @@ def test_remove_columns_invalid_names_messages():
+    t2 = Table.read(test_file)
+    assert np.all(t2['a'] == [1, 2, 3])
+    # Ensure the data wasn't written to the literal tilde-prefixed path
+    assert not os.path.exists(test_file)
```
This one blows my mind a little. So, `os.path.exists` is immune to `monkeypatch` fixture thingy?

svank replied:
The entire `os` module doesn't ever expand `~` by default (afaik), nor does the builtin `open()`, which is why `expanduser` has to be called explicitly by us

svank replied:
So that is to say, `os.path.exists` sees `~/file.csv` as a relative path containing a directory whose name is a literal `~`, and the monkey-patched environment variables never factor in

pllim commented on astropy/io/misc/tests/test_hdf5.py:
```diff
@@ -823,3 +838,17 @@ def test_overwrite_serialized_meta():
+    t1 = Table()
+    t1.add_column(Column(name='a', data=[1, 2, 3]))
+
+    t1.write(test_file, path='tilde')
```
What is the `path='tilde'` for?

taldcroft replied:
Confusingly in this context, `path` is the argument for the HDF5 writer to specify the name of the table within the HDF5 structure.

svank replied:
Indeed, multiple overloaded words in this line. I'll change the value given for `path` so that `path='tilde'` doesn't look like it might be an option flag enabling tilde-expansion

taldcroft commented on astropy/table/tests/test_table.py:
```diff
@@ -3023,3 +3038,14 @@ def test_remove_columns_invalid_names_messages():
+
+
+def test_read_write_tilde_path(home_is_tmpdir):
+    test_file = os.path.join('~', 'test.csv')
```
Need to test this for `pathlib.Path` as well. (I'm sure it will work but still...).

svank replied:
Good point, and I may need to change this in the PRs for the other submodules as well

taldcroft commented on astropy/io/misc/tests/test_hdf5.py:
```diff
@@ -823,3 +838,17 @@ def test_overwrite_serialized_meta():
         assert t3.info(out=None) == t2.info(out=None)
+
+
+@pytest.mark.skipif('not HAS_H5PY')
```
This update should work for every Table file format, so you can just write to `ascii.basic` format which has no dependence.

svank replied:
This test was to exercise HDF5 support in particular. I think my idea was that since HDF5 IO is outsourced to h5py whereas (most of?) the other table formats are handled directly within astropy, this is testing that the tilde-path expansion happens before things are passed to h5py and serves as a regression test in case the current code structure is ever changed. Otherwise this test is basically the same as the one in `test_table.py` and I can drop this one if you still think it's redundant.

taldcroft replied:
OK, good point.

taldcroft commented on astropy/io/misc/tests/test_hdf5.py:
```diff
@@ -823,3 +838,17 @@ def test_overwrite_serialized_meta():
+def test_read_write_tilde_path(tmpdir, home_is_tmpdir):
+    test_file = os.path.join('~', 'test.hdf5')
+    t1 = Table()
+    t1.add_column(Column(name='a', data=[1, 2, 3]))
```
Simpler/preferred idiom is:
```
t1['a'] = [1, 2, 3]
```

taldcroft commented on astropy/io/registry/core.py:
```diff
@@ -332,6 +339,12 @@ def write(self, data, *args, format=None, **kwargs):
 
             .. versionadded:: 4.3
         """
+        # Conservatively attempt to apply expanduser to the first argument,
```
Since this is repeating the same code, you should factor this out to a little helper function like:
```
def _expand_user_in_args(args):
    <the same code>
    return args
```
Then:
```
# Expand user (~ or ~user) if applicable in args[0]
args = _expand_user_in_args(args)
```

taldcroft commented on docs/changes/table/13129.feature.rst:
```diff
@@ -0,0 +1 @@
+Added support for reading and writing files where the file path is of the form ``~/data/file.csv``, where ``~`` indicates the home directory.
```
You need to make it clear that this applies the Unified I/O layer and not (for instance) to direct I/O within the `fits` subpackage. Note that NDData and Cosmology also support I/O through this interface and should likewise now allow `~`.

Also, this should make note that `~user/data/file.csv` also works.

Finally, it makes it a little easier to read/review if the text here is wrapped at 80  columns.

taldcroft reviewed (changes requested):
Looks nice! I have a few changes requested. In addition, you should update the Unified I/O narrative docs somewhere to indicate this enhancement.

You might also want to write similar change log fragments for `nddata`, `cosmology`, and `io.registry`. This way people interested in changes for that subpackage will see it.

svank commented:
Thanks, @taldcroft , I think I've addressed all your comments (and can squash the commit before merging).

I didn't realize `nddata` goes through `table` before reaching `io.fits` for FITS files! In the original PR that I split this off of, I have a [commit](https://github.com/astropy/astropy/pull/13107/commits/3f9e5c40b03a9618fb46f03123369e700b09dd9b) adding a test in `nddata` to exercise `CCDData` with tilde paths, but I had just assumed that would depend on the `io.fits` changes. Do you think that commit belongs in this PR?

taldcroft reviewed (approved):
Looks good, thanks for this useful feature!
